### ISSUE TITLE
Compiler crashed with 'Calling a function with a bad signature!'

### ISSUE THREAD

**Issue by jamii** [rust-lang/rust#22874]
``` rust
#![feature(hash)] // TODO is this a good idea?
#![allow(dead_code)]

#[derive(PartialOrd, PartialEq, Clone, Hash, Debug)]
enum Value {
    Least,
    String(String),
    Greatest
}

type Row = [Value];

struct Table {
    rows: [Row]
}

trait Skippable {
    fn next(&self, row: &Row, inclusive: bool) -> &Row;
}

impl Skippable for Table {
    fn next(&self, row: &Row, inclusive: bool) -> &Row {
        &self.rows[0]
    }
}

fn main() {
}
```

``` bash
amie@wanderer:~/make_you_a_database/examples$ cargo build --verbose
   Compiling make-you-a-database v0.0.1 (file:///home/jamie/make_you_a_database/examples)
     Running `rustc src/main.rs --crate-name make-you-a-database --crate-type bin -g --out-dir /home/jamie/make_you_a_database/examples/target --emit=dep-info,link -L dependency=/home/jamie/make_you_a_database/examples/target -L dependency=/home/jamie/make_you_a_database/examples/target/deps`
src/main.rs:1:12: 1:16 warning: unused or unknown feature, #[warn(unused_features)] on by default
src/main.rs:1 #![feature(hash)] // TODO is this a good idea?
                         ^~~~
src/main.rs:22:17: 22:20 warning: unused variable: `row`, #[warn(unused_variables)] on by default
src/main.rs:22  fn next(&self, row: &Row, inclusive: bool) -> &Row {
                               ^~~
src/main.rs:22:28: 22:37 warning: unused variable: `inclusive`, #[warn(unused_variables)] on by default
src/main.rs:22  fn next(&self, row: &Row, inclusive: bool) -> &Row {
                                          ^~~~~~~~~
rustc: /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/llvm/lib/IR/Instructions.cpp:281: void llvm::CallInst::init(llvm::Value*, llvm::ArrayRef<llvm::Value*>, const llvm::Twine&): Assertion `(i >= FTy->getNumParams() || FTy->getParamType(i) == Args[i]->getType()) && "Calling a function with a bad signature!"' failed.
Could not compile `make-you-a-database`.

Caused by:
  Process didn't exit successfully: `rustc src/main.rs --crate-name make-you-a-database --crate-type bin -g --out-dir /home/jamie/make_you_a_database/examples/target --emit=dep-info,link -L dependency=/home/jamie/make_you_a_database/examples/target -L dependency=/home/jamie/make_you_a_database/examples/target/deps` (status=6)
```

**Comment by zofrex** [comment]
I've made an attempt at removing as much code that doesn't affect the outcome as possible:

```rust
#![allow(dead_code)]

enum Value {
    String(String)
}

type Row = [Value];

struct Table {
    rows: [Row]
}

trait Skippable {
    fn next(&self) -> &Row;
}

impl Skippable for Table {
    fn next(&self) -> &Row {
        &self.rows[0]
    }
}

fn main() {
}
```

```
Assertion failed: ((i >= FTy->getNumParams() || FTy->getParamType(i) == Args[i]->getType()) && "Calling a function with a bad signature!"), function init, file /Users/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-mac/build/src/llvm/lib/IR/Instructions.cpp, line 281.
Abort trap: 6
```

**Comment by frewsxcv** [comment]
I reduced this even further:

```rust
#![allow(dead_code)]

enum Value {
    A(String)
}

struct Table {
    rows: [[Value]]
}

impl Table {
    fn next(&self) {
        &self.rows[0];
    }
}

fn main() {}
```

This is still an issue with Rust 1.0 beta 2

```
~/Downloads $ rustc test.rs
Call parameter type does not match function signature!
  %6 = phi %Value* [ %3, %entry-block ], [ %8, %iter_vec_loop_body ]
 { %Value*, i64 }*  call void @"_ZN16_$u5b$Value$u5d$8drop.95817h5f8aee98e1349761E"(%Value* %6)
LLVM ERROR: Broken function found, compilation aborted!

~/Downloads $ rustc --version
rustc 1.0.0-beta.2 (e9080ec39 2015-04-16) (built 2015-04-16)
```

**Comment by murarth** [comment]
I was able to get the same failed assertion with a much smaller piece of code:
```rust
trait Foo {}
fn main() {
    let _a: [Foo; 0];
}
```

```
rustc: /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/llvm/lib/IR/Instructions.cpp:281: void llvm::CallInst::init(llvm::Value*, llvm::ArrayRef<llvm::Value*>, const llvm::Twine&): Assertion `(i >= FTy->getNumParams() || FTy->getParamType(i) == Args[i]->getType()) && "Calling a function with a bad signature!"' failed.
```

**Comment by theemathas** [comment]
Related to #21748

**Comment by excaliburHisSheath** [comment]
This bug is still happening in the 10/18/2015 nightly build.

**Comment by pmarcelll** [comment]
This seems to be fixed.
@steveklabnik this can be removed from Glacier, it's just a regular error now.

**Comment by petrochenkov** [comment]
Fixed in Rust 1.2

Minimized:
```rust
struct Table {
    rows: [[String]]
}

fn f(table: &Table) -> &[String] {
    &table.rows[0]
}

fn main() {}
```

Needs test.

**Comment by mattico** [comment]
This can be closed now that #40296 is merged.
